jupyterlab/loader.js

```diff
@@ -0,0 +1,227 @@
+// Distributed under the terms of the Modified BSD License.
+
+var semver = require('semver');
+
+/**
+ * The registered module cache.
+ */
+var registered = {};
+
+/**
+ * The installed module cache.
+ */
+var installedModules = {};
+
+/**
+ * The module name lookup cache.
+ */
+var lookupCache = {};
+
+/**
+ * A cache of objects for parsed version-mangled module names.
+ */
+var nameCache = {};
+
+/**
+ * Object to store loaded and loading bundles.
+ * "0" means "already loaded".
+ * Array means "loading", array contains callbacks.
+ */
+var installedBundles = {};
+
+
+/**
+ * Define a module that can be synchronously required.
+ *
+ * @param name - The version-mangled fully qualified name of the module.
+ *   For example, "foo@1.0.1/lib/bar/baz.js".
+ *
+ * @param callback - The callback function for invoking the module.
+ */
+function defineModule(name, callback) {
+  if (!(name in registered)) {
+    registered[name] = callback;
+  }
+}
+
+
+/**
+ * Synchronously require a module that has already been loaded.
+ *
+ * @param name - The semver-mangled fully qualified name of the module.
+ *   For example, "foo@^1.1.0/lib/bar/baz.js".
+ *
+ * @returns The exports of the requested module, if registered.  The module
+ *   selected is the registered module that maximally satisfies the semver
+ *   range of the request.
+ */
+function requireModule(name) {
+  // Check if module is in cache.
+  var moduleId = findModuleId(name);
+  if(installedModules[moduleId]) {
+    return installedModules[moduleId].exports;
+  }
+
+  // Create a new module (and put it into the cache).
+  var module = installedModules[moduleId] = {
+    exports: {},
+    id: moduleId,
+    loaded: false
+  };
+
+  // Execute the module function.
+  registered[moduleId].call(module.exports, module, module.exports, requireModule);
+
+  // Flag the module as loaded.
+  module.loaded = true;
+
+  // Return the exports of the module.
+  return module.exports;
+}
+
+
+/**
+ * Ensure a bundle is loaded on a page.
+ *
+ * @param path - The public path of the bundle (e.g. "lab/jupyter.bundle.js").
+ *
+ * @param callback - The callback invoked when the bundle has loaded.
+ */
+function ensureBundle(path, callback) {
+  // "0" is the signal for "already loaded"
+  if (installedBundles[path] === 0) {
+    return callback.call(null, requireModule);
+  }
+
+  // An array means "currently loading".
+  if (Array.isArray(installedBundles[path])) {
+    installedBundles[path].push(callback);
+    return;
+  }
+
+  // Start bundle loading.
+  installedBundles[path] = [callback];
+  var head = document.getElementsByTagName('head')[0];
+  var script = document.createElement('script');
+  script.type = 'text/javascript';
+  script.charset = 'utf-8';
+  script.async = true;
+  script.onload = function() {
+    var callbacks = installedBundles[path];
+    while(callbacks.length)
+      callbacks.shift().call(null, requireModule);
+    installedBundles[path] = 0;
+  }
+  head.appendChild(script);
```

Comment:
How are errors handled? Will it be logged by default? Perhaps `script.onerror` should probably call something, or do something. I suppose the callback won't ever be called, so that's good.

It appears that node does not have error handling in its ensure function. https://www.npmjs.com/package/node-ensure has the callback taking an error parameter, but that breaks the api.